cmd/state-exec/state-exec.zig

```diff
@@ -0,0 +1,112 @@
+    ChildProcSpawn,
+};
+
+pub fn main() !void {
+    const stderr = io.getStdErr().writer();
+
+    run(stderr) catch |err| {
+        try stderr.print("{s}: ", .{execName});
+
+        switch (err) {
+            Error.ArgIterator => try stderr.print("Cannot process args.\n", .{}),
+            Error.ArgInvalidOne, Error.ArgMissingOne => try stderr.print("First arg should be a socket file.\n", .{}),
+            Error.ArgInvalidTwo, Error.ArgMissingTwo => try stderr.print("Second arg should be a language runtime.\n", .{}),
+            Error.ArgCollector => try stderr.print("Cannot setup arg collector.\n", .{}),
+            Error.ArgCollectRunt => try stderr.print("Cannot collect runtime arg.\n", .{}),
+            Error.ArgCollectUsr => try stderr.print("Cannot collect user args.\n", .{}),
+            Error.InspectSelfPath => try stderr.print("Cannot obtain path to this executable.\n", .{}),
+            Error.ThreadSpawn => try stderr.print("Cannot spawn thread for heartbeat.\n", .{}),
+            Error.ChildProcInit => try stderr.print("Cannot initialize child process for runtime.\n", .{}),
+            Error.ChildProcSpawn => try stderr.print("Cannot spawn child process for runtime.\n", .{}),
+        }
+
+        try stderr.print("{s}: This application is not intended to be user serviceable; Please contact support for assistance.\n", .{execName});
+
+        process.exit(1);
+    };
+}
+
+fn sendMsgToServer(stderr: fs.File.Writer, a: mem.Allocator, path: []const u8, pid: i32, exec: []const u8) !void {
+    const clientMsgFmt = "heart<{d}<{s}";
+
+    const conn = net.connectUnixSocket(path) catch |err| {
+        try stderr.print("{s}: Cannot connect to socket: {s}.\n", .{ execName, err });
+        return;
+    };
+    defer conn.close();
+
+    var clientMsg = try fmt.allocPrint(a, clientMsgFmt, .{ pid, exec });
+    _ = conn.write(clientMsg) catch |err| {
+        try stderr.print("{s}: Cannot write to socket connection: {s}.\n", .{ execName, err });
+        return;
+    };
+
+    var buf: [1024]u8 = undefined;
+    _ = conn.read(buf[0..]) catch |err| {
+        try stderr.print("{s}: Cannot read from socket connection: {s}.\n", .{ execName, err });
+        return;
+    };
+}
+
+fn run(stderr: fs.File.Writer) Error!void {
+    var arena = heap.ArenaAllocator.init(heap.page_allocator);
+    defer arena.deinit();
+    const a = arena.allocator();
+
+    var argIt = process.argsWithAllocator(a) catch return Error.ArgIterator;
+    defer argIt.deinit();
+
+    _ = argIt.skip();
+    const path = (argIt.next(a) orelse return Error.ArgMissingOne) catch return Error.ArgInvalidOne;
+    const runt = (argIt.next(a) orelse return Error.ArgMissingTwo) catch return Error.ArgInvalidTwo;
+
+    var pid: i32 = @truncate(i32, @bitCast(i64, Thread.getCurrentId()));
+
+    const exec = fs.selfExePathAlloc(a) catch return Error.InspectSelfPath;
+
+    const clientThread = Thread.spawn(.{}, sendMsgToServer, .{ stderr, a, path, pid, exec }) catch {
```

Comment:
I'll make a quick example to prove this in the morning, but, in theory, the heartbeat signal is being sent concurrently while the runtime is launched, so the user should not notice delay in their runtime usage start. The duration of that thread should be very short since it's using a very fast form of communication with the service, so there should be little to no delay in runtime duration.

Reply:
This is assuming that our socket performs faster than the actual executable that's being ran underneath. Seems plausible but yes; please verify (maybe make a very lightweight hello world executable to test with).

Reply:
On my machine...
Calling the thread that calls the service and not waiting for it takes .08ms.
Calling the thread that calls the service and waiting for it takes .10ms.
Calling the thread that calls the service and putting it to the side while python3 is called as a child process takes around 12.80 - 13.00ms.
Calling python3 directly takes around 12.06-12.28ms.

Reply:
If you were testing with Python3 then that invalidates the test. Python3 is just ONE potential executable that the executor could invoke. A proper test should use a very lightweight executable such that the overhead of the executable itself is as minimal as possible and we can properly assert that the threading approach doesn't cause any significant delays.

Reply:
I might be misunderstanding... The cost of the concurrent call  seems to be .08ms. The cost of the heartbeat messaging seems to be .02ms. So, .10ms total. It's probably safe to assume all runtimes will take longer than that. Where this helps us is if the service has some delay responding and that .10ms becomes 20ms... the user will see their output immediately, but the process won't complete until the end of 20ms. If we run this serially, then that 20ms will need to pass before the user sees any output from their runtime.